crates/containerd-shim-wasmtime/src/instance.rs

```diff
@@ -309,3 +350,146 @@ fn wasi_builder(
         .preopened_dir("/", "/", dir_perms, file_perms)?;
     Ok(builder)
 }
+
+fn wasi_store_for_request(engine: &wasmtime::Engine, req_id: u64) -> Store<WasiCtx> {
+    let mut builder = wasi_preview2::WasiCtxBuilder::new();
+
+    // TODO: populate envs from RuntimeContext?
+    builder.env("REQUEST_ID", req_id.to_string());
+
+    let ctx = WasiCtx {
+        wasi_preview1: builder.build_p1(),
+        wasi_preview2: builder.build(),
+        wasi_http: WasiHttpCtx::new(),
+        resource_table: ResourceTable::default(),
+    };
+
+    Store::new(engine, ctx)
+}
+
+struct ProxyHandlerInner {
+    instance_pre: ProxyPre<WasiCtx>,
+    next_id: AtomicU64,
+}
+
+impl ProxyHandlerInner {
+    fn next_req_id(&self) -> u64 {
+        self.next_id.fetch_add(1, Ordering::Relaxed)
+    }
+}
+
+#[derive(Clone)]
+struct ProxyHandler(Arc<ProxyHandlerInner>);
+
+impl ProxyHandler {
+    fn new(instance_pre: ProxyPre<WasiCtx>) -> Self {
+        Self(Arc::new(ProxyHandlerInner {
+            instance_pre,
+            next_id: AtomicU64::from(0),
+        }))
+    }
+}
+
+type Request = hyper::Request<hyper::body::Incoming>;
+
+async fn serve_conn(addr: std::net::SocketAddr, instance: ProxyPre<WasiCtx>) -> Result<()> {
+    use hyper::server::conn::http1;
+
+    let socket = match addr {
+        SocketAddr::V4(_) => tokio::net::TcpSocket::new_v4()?,
+        SocketAddr::V6(_) => tokio::net::TcpSocket::new_v6()?,
+    };
+    // Conditionally enable `SO_REUSEADDR` depending on the current
+    // platform. On Unix we want this to be able to rebind an address in
+    // the `TIME_WAIT` state which can happen then a server is killed with
+    // active TCP connections and then restarted. On Windows though if
+    // `SO_REUSEADDR` is specified then it enables multiple applications to
+    // bind the port at the same time which is not something we want. Hence
+    // this is conditionally set based on the platform (and deviates from
+    // Tokio's default from always-on).
+    socket.set_reuseaddr(!cfg!(windows))?;
+    socket.bind(addr)?;
+
+    let listener = socket.listen(100)?;
```

Comment:
question: perhaps the `backlog` argument to `listen()` could come from RuntimeContext? If it's unspecified, then default it to `100`?